browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc

```diff
@@ -53,12 +69,111 @@ using mojom::CharacterType;
 using mojom::ConversationTurn;
 using mojom::ConversationTurnVisibility;
 
+#if BUILDFLAG(ENABLE_PRINT_PREVIEW)
+class PreviewPageTextExtractor {
+ public:
+  PreviewPageTextExtractor(base::ReadOnlySharedMemoryRegion pdf_region,
+                           base::OnceCallback<void(std::string)> callback,
+                           uint32_t max_page_content_length,
+                           std::optional<bool> pdf_use_skia_renderer_enabled)
+      : pdf_region_(std::move(pdf_region)),
+        callback_(std::move(callback)),
+        max_page_content_length_(max_page_content_length) {
+    DCHECK(!pdf_to_bitmap_converter_.is_bound());
+    GetPrintingService()->BindPdfToBitmapConverter(
+        pdf_to_bitmap_converter_.BindNewPipeAndPassReceiver());
+    pdf_to_bitmap_converter_.set_disconnect_handler(
+        base::BindOnce(&PreviewPageTextExtractor::BitmapConverterDisconnected,
+                       base::Unretained(this)));
+    if (pdf_use_skia_renderer_enabled.has_value()) {
+      pdf_to_bitmap_converter_->SetUseSkiaRendererPolicy(
+          pdf_use_skia_renderer_enabled.value());
+    }
+  }
+
+  void StartExtract() {
+    pdf_to_bitmap_converter_->GetPdfPageCount(
+        pdf_region_.Duplicate(),
+        base::BindOnce(&PreviewPageTextExtractor::OnGetPageCount,
+                       base::Unretained(this)));
+  }
+
+  void OnGetPageCount(std::optional<uint32_t> page_count) {
+    if (!page_count.has_value() || !page_count.value()) {
+      std::move(callback_).Run("");
+      return;
+    }
+    total_page_count_ = page_count.value();
+    pdf_to_bitmap_converter_->GetBitmap(
```

Comment:
I'd refactor this a bit to move `GetBitmap` call into a separate method and call it both from `OnGetPageCount` and `OnGetTextFromImage`.

something like `ScheduleNextPageOrComplete`.

Reply:
refactored in 1a018f1502b72df3f1b6a6e1f044c3f8bd9bff17